pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -966,35 +977,76 @@ func (t *Test) choosePreviousReleases() ([]*clusterupgrade.Version, error) {
 		return []*clusterupgrade.Version{pred, predPred}, nil
 	}
 
+	// possibleUpgradePathsMap maps a version to the possible upgrade paths that
+	// can be taken with exactly n upgrades.
+	possibleUpgradePathsMap := make(map[*clusterupgrade.Version]map[int][][]*clusterupgrade.Version)
+	// findPossibleUpgradePaths finds all legal upgrade paths ending at version `v` with
+	// exactly `numUpgrades` upgrades.
+	var findPossibleUpgradePaths func(v *clusterupgrade.Version, numUpgrades int) ([][]*clusterupgrade.Version, error)
+	findPossibleUpgradePaths = func(v *clusterupgrade.Version, numUpgrades int) ([][]*clusterupgrade.Version, error) {
+		// If there are no upgrades, then the only possible path is the current version.
+		if numUpgrades == 0 {
+			return [][]*clusterupgrade.Version{{v}}, nil
+		}
+		// Check if we have already computed the possible upgrade paths for this version.
+		if _, ok := possibleUpgradePathsMap[v]; !ok {
+			possibleUpgradePathsMap[v] = make(map[int][][]*clusterupgrade.Version)
+		} else if paths, ok := possibleUpgradePathsMap[v][numUpgrades]; ok {
+			return paths, nil
+		}
+
+		// The possible upgrade paths for this version with exactly `numUpgrades` upgrades
+		// is the possible upgrade paths of `v's` predecessors with `numUpgrades-1`
+		// upgrades plus version `v`.
+		var paths [][]*clusterupgrade.Version
+		predecessors, err := possiblePredecessorsFor(v)
+		if err != nil {
+			return nil, err
+		}
+		for _, pred := range predecessors {
+			// If the predecessor is older than the minimum bootstrapped version, then
+			// it's not a legal upgrade path.
+			if t.options.minimumBootstrapVersion != nil && !pred.AtLeast(t.options.minimumBootstrapVersion) {
```

Comment:
`!x.AtLeast(y)` implies `x < y`. Maybe it's worth adding `Version.Less`? :)

Reply:
Added as a helper in `clusterupgrade`. I could add it in the `version` pkg too, but I vaguely remember seeing some discussion around why we don't want to add something like that. Can't find it though, maybe it rings a bell for you?